File: cpp/LargestNumber.cc

class Solution {
public:
    string largestNumber(vector<int> &num) {
        vector<string> strs;
        for (auto i: num) {
            strs.push_back(to_string(i));
        }
        
        sort(strs.begin(), strs.end(), [](const string& s1, const string& s2) { return s1 + s2 > s2 + s1;}); // note we want desc order
        string res = "";
        int i = 0; // ignore leading 0s
        while (i < strs.size() && strs[i] == "0") {
            i++;
        }
        
        // if all "0", ...
        if (i == strs.size()) {
            return "0";
        }
        
        for (; i < strs.size(); i++) {
            res += strs[i];
        }
    
        return res;
    }
};

File: java/src/main/java/org/speedfirst/leetcode/string/LargestNumber.java
package org.speedfirst.leetcode.string;

import java.util.ArrayList;
import java.util.List;

/**
 * @author jiankuan
 * @since 1/15/15
 */
public class LargestNumber {
    public String largestNumber(int[] num) {
        List<String> strs = new ArrayList<>();
        for (int n: num) {
            strs.add(String.valueOf(n));
        }

        strs.sort((String s1, String s2) -> (s2 + s1).compareTo(s1 + s2));

        int i = 0;
        while (i < strs.size() && strs.get(i).equals("0")) {
            i++;
        }

        if (i == strs.size()) {
            return "0";
        }

        StringBuilder sb = new StringBuilder();
        while (i < strs.size()) {
            sb.append(strs.get(i));
            i++;
        }

        return sb.toString();
    }

    public static void main(String[] args) {
        int[] num = new int[]{1};
        String res = new LargestNumber().largestNumber(num);
        System.out.println(res);
    }
}